R/trackhub.R

```diff
@@ -0,0 +1,301 @@
+    if (create) {
+        if (uriExists(uri)) {
+            message("NOTE: '", uri, "' already exists")
+            create <- FALSE
+        } ## must create this before calling normURI (requires existence)
+        else createResource(uri, dir = TRUE)
+    }
+    th <- new("TrackHub", uri = normURI(uri))
+    if (create)
+        createResource(hubFile(th))
+    th
+}
+
+setAs("character", "TrackHub", function(from) TrackHub(from))
+
+### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+### TrackHubGenome class
+###
+
+setClass("TrackHubGenome",
+         representation(trackhub = "TrackHub",
+                        genome = "character"),
+         contains = "TrackDb")
+
+trackhub <- function(x, ...) x@trackhub
+trackDbFile <- function(x,y) paste(trimSlash(uri(x)), y, sep = "/")
+twobitFile <- function(x,y) paste(trimSlash(uri(x)), y, sep = "/")
+
+getTrackDbContent <- function(x) {
+    content <- readLines(x, warn = FALSE)
+    # TODO
+    # store content in right data structure
+}
+
+createTrackHubGenome <- function(x, genomeRecord) {
+    hubContent <- getHubContent(hubFile(trackhub(x)))
+    genomesFilePath <- genomesFile(trackhub(x), hubContent["genomesFile"])
+    if (uriExists(genomesFilePath) && genome(x) %in% genome(trackhub(x))) {
+        message("NOTE: Genome '", genome(x), "' already exists")
+        return ()
+    }
+    createResource(uri(x), dir = TRUE)
+    createResource(genomesFilePath)
+    setHubContentList(genomeRecord, genomesFilePath)
+}
+
+setMethod("genome", "TrackHubGenome", function(x) x@genome)
+
+setMethod("uri", "TrackHubGenome", function(x)
+          paste(trimSlash(uri(trackhub(x))), genome(x), sep = "/"))
+
+setMethod("names", "TrackHubGenome", function(x) {
+    trackDbValue <- getGenomesKey(x, "trackDb")
+    if (!isFieldEmpty(trackDbValue)) {
+        trackDbFilePath <- trackDbFile(trackhub(x), trackDbValue)
+        getTrackDbContent(trackDbFilePath)
+    }
+    else stop("genomes.txt: 'trackDb' does not contain valid reference to trackDb file")
+})
+
+setMethod("organism", "TrackHubGenome", function(object) {
+    organism <- getGenomesKey(object, "organism")
+    as.character(organism)
+})
+
+setMethod("referenceSequence", "TrackHubGenome", function(x) {
+    twoBitPathValue <- getGenomesKey(x, "twoBitPath")
+    if (!isFieldEmpty(twoBitPathValue)) {
+        twoBitFilePath <- twobitFile(trackhub(x), twoBitPathValue)
+        import(twoBitFilePath)
+    }
+    else message("genome.txt: 'twoBitPath' does not contain a reference to a file")
+})
+
+setReplaceMethod("referenceSequence", "TrackHubGenome", function(x, value) {
+    twoBitPathValue <- getGenomesKey(x, "twoBitPath")
+    if (!isFieldEmpty(twoBitPathValue)) {
+        twoBitFilePath <- twobitFile(trackhub(x), twoBitPathValue)
+        export.2bit(value, twoBitFilePath)
+        x
+    }
+    else message("genome.txt: 'twoBitPath' does not contain a reference to a file")
```

Comment:
Not sure why these are `message()` calls instead of `stop()` calls.

Reply:
Yes, here it should call to `stop()`.